crates/wasi-common/yanix/src/file.rs

```diff
@@ -165,34 +207,47 @@ pub unsafe fn openat<P: AsRef<Path>>(
     mode: Mode,
 ) -> Result<RawFd> {
     let path = cstr(path)?;
-    from_result(libc::openat(
-        dirfd,
+    let fd: libc::c_int = from_result(libc_openat(
+        dirfd as libc::c_int,
         path.as_ptr(),
         oflag.bits(),
         libc::c_uint::from(mode.bits()),
-    ))
+    ))?;
+    Ok(fd as RawFd)
 }
 
 pub unsafe fn readlinkat<P: AsRef<Path>>(dirfd: RawFd, path: P) -> Result<OsString> {
     let path = cstr(path)?;
-    let buffer = &mut [0u8; libc::PATH_MAX as usize + 1];
-    let nread = from_result(libc::readlinkat(
-        dirfd,
-        path.as_ptr(),
-        buffer.as_mut_ptr() as *mut _,
-        buffer.len(),
-    ))?;
-    // We can just unwrap() this, because readlinkat returns an ssize_t which is either -1
-    // (handled above) or non-negative and will fit in a size_t/usize, which is what we're
-    // converting it to here.
-    let nread = nread.try_into().unwrap();
-    let link = OsStr::from_bytes(&buffer[0..nread]);
-    Ok(link.into())
+    // Start with a buffer big enough for the vast majority of paths.
+    let mut buffer = Vec::with_capacity(256);
+    loop {
```

Comment:
For WASI, I want to avoid PATH_MAX because the actual value of PATH_MAX depends on the underlying host OS, so it's better for WASI code to avoid making size assumptions. But you're right that other platforms don't need this. I've now made this code conditional to restore stack allocation.